doc/barclamp_development_exercises.md

```diff
@@ -0,0 +1,341 @@
+repeat the action. Chef allows us to manage this by setting persistent flags
+on the node that we can evaluate before deciding to include a resource in the
+Chef run.
+
+To demonstrate the importance of knowing the difference between the compile and
+converge phases, let's first do this incorrectly.
+
+First, undo the result of the last successful ``mapping_populate`` run by
+emptying the table:
+
+```
+# mysql keystone -e 'delete from id_mapping'
+```
+
+Also, change the command in the ``execute`` resource you added in the previous
+exercise to return an error, for example by changing the command to ``false``.
+This will illustrate what happens when the command fails to execute.
+
+After the ``execute`` resource in the recipe, set a flag on the node object to
+declare that the mapping population has been run, for example:
+
+```
+node.set[:keystone][:mapping_populated] = true
+node.save
+```
+
+Also add an ``only_if`` parameter to the ``execute`` resource so that it will
+only run if that flag is not set. The (naive) idea is that the ``execute``
+resource should run once, then the flag should be set, and then on the next Chef
+run it won't be executed. Moreover, if the resource fails, we would want
+execution to stop so that the problem can be corrected, and then we would want
+the next run to try the mapping population again.
+
+Upload the cookbook and reapply the proposal. Notice the proposal does not fail
+on the invalid command in the ``execute`` resource. Also check the database on
+the controller and notice that nothing happened.
+
+Fix the command in the ``execute`` resource to use the correct mapping populate
+command. Upload the cookbook again and reapply the proposal. This should again
+be successful, but the command still will have not populated the database.
+
+Why not? Because the code to set the ``:mapping_populated`` flag was run in
+the *compile* phase, long before the ``execute`` resource was run. When it came
+time to run the resource, it saw that the flag on the node was already set and
+decided not to run the resource. You can examine the node attributes with knife:
+
+```
+# knife node list
+# ...
+# knife node edit <controller1>
+```
+
+Go ahead and reset the node by deleting that ``"mapping_populated"`` flag in the
+node JSON in knife. Do the same on all other controllers.
+
+Now let's do it correctly. In the recipe, wrap the node attribute setting in a
+``ruby_block``:
+
+```
+ruby_block "mark node for keystone mapping_populate" do
+  block do
+    node.set[:keystone][:mapping_populated] = true
+    node.save
+  end
+  action :nothing
+  subscribes :create, "execute[keystone-manage mapping_populate]", :immediately
+end
+```
+
+Make sure the ``ruby_block`` comes before the closing sync mark. The
+``ruby_block`` is a Chef resource, which means it won't be executed until the
+*converge* phase. Now when you upload the cookbook and apply the proposal, the
```

Comment:
.. or to put it differnetly, it actually will get run twice: - during compile phase the "block" is being expanded and the expansion is saved (so input parameters are copied and "frozen at that time" and then during the -convergence phase the list of compiled resources are being executed, potentially triggering from the list of resources additional runs (via subscribe/trigger logic)

Reply:
Rephrased this and expanded on it in the intro for this section